File: leetcode/go/3030.find-the-grid-of-region-average.go
func resultGrid(image [][]int, threshold int) [][]int {
     m, n := len(image), len(image[0])

    sum := make([][]int, m)
    counts := make([][]int, m)
    for i := range sum {
        sum[i] = make([]int, n)
        counts[i] = make([]int, n)
    }

    is_valid_region := func(i, j int) bool {
        for x := i; x < i+3; x++ {
            for y := j; y < j+3; y++ {
                if x > i && abs(image[x][y]-image[x-1][y]) > threshold {
                    return false
                }
                if y > j && abs(image[x][y]-image[x][y-1]) > threshold {
                    return false
                }
            }
        }
        return true
    }

    for i := 0; i < m-2; i++ {
        for j := 0; j < n-2; j++ {
            if is_valid_region(i, j) {
                subgrid_sum := 0
                for x := i; x < i+3; x++ {
                    for y := j; y < j+3; y++ {
                        subgrid_sum += image[x][y]
                    }
                }
                subgrid_sum /= 9
                for x := i; x < i+3; x++ {
                    for y := j; y < j+3; y++ {
                        sum[x][y] += subgrid_sum
                        counts[x][y] += 1
                    }
                }
            }
        }
    }

    for i := 0; i < m; i++ {
        for j := 0; j < n; j++ {
            if counts[i][j] > 0 {
                image[i][j] = sum[i][j] / counts[i][j]
            }
        }
    }

    return image
}

func abs(x int) int {
    if x < 0 {
        return -x
    }
    return x
}

File: leetcode/cpp/3030.find-the-grid-of-region-average.cpp
class Solution {
public:
    vector<vector<int>> resultGrid(vector<vector<int>>& image, int threshold) {
        int m = image.size(), n = image[0].size();

        std::vector<std::vector<int>> sum(m, std::vector<int>(n, 0));
        std::vector<std::vector<int>> counts(m, std::vector<int>(n, 0));

        auto is_valid_region = [&image, threshold](int i, int j) -> bool {
            for (int x = i; x < i + 3; ++x) {
                for (int y = j; y < j + 3; ++y) {
                    if (x > i && std::abs(image[x][y] - image[x - 1][y]) > threshold) {
                        return false;
                    }
                    if (y > j && std::abs(image[x][y] - image[x][y - 1]) > threshold) {
                        return false;
                    }
                }
            }
            return true;
        };

        for (int i = 0; i < m - 2; ++i) {
            for (int j = 0; j < n - 2; ++j) {
                if (is_valid_region(i, j)) {
                    int subgrid_sum = 0;
                    for (int x = i; x < i + 3; ++x) {
                        for (int y = j; y < j + 3; ++y) {
                            subgrid_sum += image[x][y];
                        }
                    }
                    subgrid_sum /= 9;
                    for (int x = i; x < i + 3; ++x) {
                        for (int y = j; y < j + 3; ++y) {
                            sum[x][y] += subgrid_sum;
                            counts[x][y] += 1;
                        }
                    }
                }
            }
        }

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (counts[i][j] > 0) {
                    image[i][j] = sum[i][j] / counts[i][j];
                }
            }
        }

        return image;
    }
};

